### PR TITLE
Fix incorrect pattern warning "unreachable pattern"

### LINKED ISSUE #70372: Incorrect "unreachable pattern" warning.
Not much to comment about. The supposedly “unreachable” pattern is obviously reached.
```rust

fn main() {
    match (3,42) {
        (a,_) | (_,a) if a > 10 => {println!("{}", a)}
        _ => ()
    }
}

```

([Playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=3f11f5a60aaa9b7c478583c18becb57e))

Output:

```
42

```

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning: unreachable pattern
 --> src/main.rs:4:17
  |
4 |         (a,_) | (_,a) if a > 10 => {println!("{}", a)}
  |                 ^^^^^
  |
  = note: `#[warn(unreachable_patterns)]` on by default

    Finished release [optimized] target(s) in 0.78s
     Running `target/release/playground`

```

This issue has been assigned to @AminArria via [this comment](https://github.com/rust-lang/rust/issues/70372#issuecomment-604096011).

### PR DESCRIPTION
**Pull request by AminArria**
Fixes #70372

Added `is_under_guard` parameter to `_match::is_useful` and only add it to the matrix if `false`

Tested with:
```rust
#![feature(or_patterns)]
fn main() {
    match (3,42) {
        (a,_) | (_,a) if a > 10 => {println!("{}", a)}
        _ => ()
    }

    match Some((3,42)) {
        Some((a, _)) | Some((_, a)) if a > 10 => {println!("{}", a)}
        _ => ()
    
    }

    match Some((3,42)) {
        Some((a, _) | (_, a)) if a > 10 => {println!("{}", a)}
        _ => ()
    }
}
```

### CODE REVIEW

File: src/librustc_mir_build/hair/pattern/check_match.rs [hunk 1]
```diff
@@ -410,7 +410,7 @@ fn check_not_useful<'p, 'tcx>(
 ) -> Result<(), Vec<super::Pat<'tcx>>> {
     let wild_pattern = cx.pattern_arena.alloc(super::Pat::wildcard_from_ty(ty));
     let v = PatStack::from_pattern(wild_pattern);
-    match is_useful(cx, matrix, &v, ConstructWitness, hir_id, true) {
```
Comment (Centril) [inline src/librustc_mir_build/hair/pattern/check_match.rs]: Can you leave a note re. the `false` in this case?

**Review by Centril — COMMENTED** [review]
Looks good. :)

Please also add the test cases mentioned in the OP in `src/test/ui/or-patterns/issue-70413-no-unreachable-pat-and-guard.rs`, with `#![deny(unreachable_patterns)]` and write `// check-pass` at the top of the file. Also make sure to `./x.py test --stage 1 --bless --pass check src/test/ui` to adjust the fallout in UI tests (see also https://rustc-dev-guide.rust-lang.org/tests/adding.html).

You'll also need to run `./x.py fmt` to pacify the rustfmt checks in CI.

**Comment by Centril** [comment]
r? @varkor
cc @Nadrieril

File: src/librustc_mir_build/hair/pattern/_match.rs [hunk 2]
```diff
@@ -1664,7 +1669,9 @@ crate fn is_useful<'p, 'tcx>(
                     bug!("Encountered or-pat in `v` during exhaustiveness checking")
                 }
             }
-            matrix.push(v);
+            if !is_under_guard {
```
Comment (Centril) [inline src/librustc_mir_build/hair/pattern/_match.rs]: Maybe add a comment here as well re. the purpose of this check.

File: src/librustc_mir_build/hair/pattern/check_match.rs (hunk 1 above)
Comment (AminArria) [reply src/librustc_mir_build/hair/pattern/check_match.rs]: Added in 4b1aa9fa

File: src/librustc_mir_build/hair/pattern/_match.rs (hunk 2 above)
Comment (AminArria) [reply src/librustc_mir_build/hair/pattern/_match.rs]: 4b1aa9fa

**Review by Nadrieril — APPROVED** [review]

**Comment by varkor** [comment]
@AminArria: could you rebase the three implementation-related commits into one? It's fine to have a separate commit for the tests. This just helps to keep the git history a little cleaner. After that, this all looks good, thanks!

**Comment by AminArria** [comment]
@varkor did the rebase

**Comment by varkor** [comment]
Thank you!

@bors r=Centril,Nadrieril,varkor rollup

**Pull request merged**